lib/spack/llnl/util/filesystem.py

```diff
@@ -956,11 +956,26 @@ class HeaderList(FileList):
     commonly used compiler flags or names.
     """
 
+    include_regex = re.compile(r'(.*)(include)(.*)')
+
     def __init__(self, files):
         super(HeaderList, self).__init__(files)
 
         self._macro_definitions = []
 
+    @property
+    def directories(self):
+        dir_list = super(HeaderList, self).directories
+        values = []
+        for d in dir_list:
+            # If the path contains a subdirectory named 'include' then stop
+            # there and don't add anything else to the path.
```

Comment:
could you please add a regression test for this logic?

Reply:
Sure, I am about to add a list of things still to be done in the description above :slightly_smiling_face:

Reply:
So if I understand: this presumes that wherever the `include/` directory exists in the hierarchy, that's what you want to include. So in the case of CMake from https://github.com/spack/spack/issues/10601 which was tripping over `libuv` which had

```
libuv-prefix/
  include/
    something.h
    uv/
      errno.h
```

This would stop at `libuv-prefix/include`. This seems like a sensible default. Generally I think the effect would be to add `<prefix>/include`, although if packages store headers in prefixes like `<prefix>/alt-include/` that would work there (I don't know of issues in Spack like that though).

The main potential issue with this is that it's difficult for users to customize the behavior of `-I` entries returned. For example I'm not sure if this resolves https://github.com/spack/spack/issues/8324, in particular if there are any header files in `/apps/intel/ComposerXE2017/include`.

Reply:
> For example I'm not sure if this resolves #8324, in particular if there are any header files in `/apps/intel/ComposerXE2017/include`.

that's very true, I completely forgot about this. But AFAIR `headers()` in `IntelPackage` have complicated but correct implementation and they won't point to the un-versioned `include` folder.

On another hand, there is `/apps/intel/ComposerXE2017/lib`, so https://github.com/spack/spack/pull/10622 most likely broke this. 

@scheibelp could we add `prefix.lib`/`prefix.lib64` only if `Libraries` are empty?

Reply:
@alalazo 

> * Implement a way for packages to opt-out from injecting include dirs into compiler wrappers (this could be done in a separate PR and refers to #8324 - thanks to @scheibelp for the reference)

given above, I don't think you need to worry about this.

Reply:
> The main potential issue with this is that it's difficult for users to customize the behavior of -I entries returned. For example I'm not sure if this resolves #8324, in particular if there are any header files in /apps/intel/ComposerXE2017/include

Assuming packages of these kind are rare we might make `HeaderList.directories` settable, and let packages that need a custom behavior construct an appropriate `HeaderList` object in their `headers` method?

Reply:
> Assuming packages of these kind are rare we might make `HeaderList.directories` settable, and let packages that need a custom behavior construct an appropriate `HeaderList` object in their `headers` method?

I would not bother until we hit this issue with some package. AFAICT Changes in this PR won’t be a problem for intel packages so long as we don’t unconditionally add prefix.include

Reply:
Considering the discussion I'll skip over that for now. If needed imo we need to implement a proper:
```
@directories.setter
def directories(self, value):
    pass
```
in the `HeaderList` class.